QEP-label-connectors.rst

```diff
@@ -0,0 +1,109 @@
+- a straight line
+- two connected straight lines, with an horizontal part and a free part
+- three connected straight lines, with two horizontal parts and a vertical part
+- a curve with one bending point
+- a curve with two bending points
+
+.. image:: connector_types.jpg
+
+To actually draw the connectors, additional properties (aside X and Y position) are needed like:
+- style of the line (plain, dotted, dashed, etc.)
+- thickness of the line
+- color of the line
+- presence of arrow heads on both ends
+- style of the arrow heads
+- size of the arrow heads (absolute or relative)
+- color of the arrow heads
+
+Placement of the two ends also need additional properties:
+- position of the ending on the side of the label : up, down, right, left, center, etc.
+- position of the ending on the side of the feature:
+  - for points: no particular option, it targets the point in itself
+  - for lines: beginning, middle, center
+  - for polygons: centroid of the polygon or on the border
+
+Certain types of connectors may also need specific properties like coordinates of the control points for curves (Bézier or B-splines). Coordinates
+of control points can only be data-defined or defined by 'null' which will make the display engine use 'best' coordinates.
+
+Implementation details
+----------------------
+
+Connectors are drawn to the Qt canvas, using the QPainter API. In order to be able to reuse the drawing code in other parts (possibly in a new symbology for lines), some
+classes will be designed.
+These classes are composed mostly of read/write properties with a central method aiming at doing the actual drawing with a render context.
+
+QgsConnectorRenderer
+
+lineStyle
+lineColor
+headStyle
+headColor
+tailStyle
+tailColor
+isCurved
+void render( point[], renderContext )
+
+In the current code base, the following methods would be modified:
+- QgsPalLabeling::drawLabeling() : extract necessary property values and data-defined values for connectors
+- new method drawConnector, similar to drawLabel that will instanciate a QgsConnectorRenderer and call render
+
+
+GUI Changes
+-----------
+
+The labeling part of the "properties" dialog of a layer will need a new entry for connector-related properties. These properties
+will be available only if X and Y positions are data-defined (i.e. manual placement).
+
+.. image:: connector_properties.png
+
+Each of the property could also be data-defined (not represented here).
+
+From the user perspective, new tools will be available to easily manipulate label connectors. To be consistent with current label manual placement toolboxes,
+new controls will be added to manually place control points for connector types that need them.
+
+Optionaly, it would be nice for the user to be able to edit per-connector properties, i.e. properties that are data-defined via a GUI dialog. This could
+be achieved by reusing the connector properties dialog and connecting it to one particular feature.
+
```

Comment:
Again, I like the general direction this is going, with the exceptions noted above. 

Have you thought about automatic creation of a connector when "display a connector" is enabled and the user first moves a label from it's automatic placement? 

Also, what considerations have been given about automatic creation of dynamic connectors if a label is further than a specified tolerance from a feature? Hopefully we'll eventually also see support for that in QGIS, so this work should be done allowing for that possibility in future.